### ISSUE TITLE
libmysql.dll 8.0.35

### ISSUE THREAD

**Issue by c-michail** [hmailserver/hmailserver#479]
Hello,
  It seems libmysql.dll 8.0.35 changed some things and doesn't work with hmailserver. (x64)

Thank you

**Comment by RvdHout** [comment]
Maybe simply don't use libmysql.dll 8.x ???

**Comment by RvdHout** [comment]
It works OK with the libmysql version available here: https://downloads.mysql.com/archives/c-c/ 

If your server is also MySQL 8 you NEED to enable the legacy option during install or execute the following query in MYSQL Workbench or something alike to make it accept ols style passwords, like:

```
ALTER USER '_root'@'localhost_' IDENTIFIED WITH mysql_native_password BY '_password_';
flush privileges;
```

**Comment by chmichael** [comment]
Yes i know, it works ok with libmysql 8.0.34 and below (with legacy option) but is it so hard to implement the new mysql auth for hmailserver ?

**Comment by SorenRR** [comment]
Good question, you already have access to the source code of hMailServer on GitHub, have a look and tell us.

**Comment by martinknafve** [comment]
I installed MySQL 8.4 and did not run into this issue. Not sure if I missed something. Latest 5.7-build also supports `libmariadb.dll` client instead of `libmysql.dll`, so guessing this is no longer an issue. @RvdHout , do you know more?

**Comment by RvdHout** [comment]
Nope, MariaDB user myself, but i think this topic was about using hMailServer with libmysql.dll 8.x against a MySQL 8.x Server instance, with the move from MySQL 5.x to MySQL 8.x Oracle switched to a new default password plugin, eg: caching_sha2_password

To keep using MySQL 5.x type formats, you could do something like:


System wide through: **my.ini**
```
[mysqld]
; Keep using MySQL 5 Passwords
default_authentication_plugin=mysql_native_password
```

Per user
```
ALTER USER 'user'@'localhost_' IDENTIFIED WITH mysql_native_password BY '_password_';
flush privileges;
```

But no idea if libmariadb.dll supports that, or has a auto fallback thingy like with TLS/SSL

**Comment by martinknafve** [comment]
I installed MySQL 8.4. I was not able to connect to it using hMailServer 5.6. With hMailServer 5.7, I can connect with:

* libmysql.dll 5.7.38
* libmysql.dll 8.0
* libmariadb.dll 3.4.9

I'll close this. Feel free to re-open if you still have connectivity issues.

**Comment by RvdHout** [comment]
As a test trying to install MySQL Community Server, https://dev.mysql.com/downloads/mysql/ (26.7.0) and with using libmariadb.dll

On database creation:
```
Starting task...
Please wait while creating database...
Plugin caching_sha2_password could not be loaded: The specified module could not be found. Library path is 'caching_sha2_password.dll'
```

> Folder Structure: When deploying your application, you must place the plugin DLL in a folder named plugin located in the same directory as your application's executable or libmariadb.dll

libmariadb.dll in `./Bin/` (libmariadb.dll 3.4.9)
caching_sha2_password.dll is in` ./Bin/plugin/`

**Comment by RvdHout** [comment]
If I install mariadb-connector-c-3.4.9-win64.msi and the repeat the steps to make a new hmailserver database I have no problem whatsoever, so that deployment instruction (above) obviously doesn't work 

I noticed below entries are added to PATH environment variables by mariadb-connector-c-3.4.9-win64.msi  though

C:\Program Files\MariaDB\MariaDB Connector C 64-bit\lib\
C:\Program Files\MariaDB\MariaDB Connector C 64-bit\lib\plugin\

**Comment by RvdHout** [comment]
Maybe it needs a mysql_options(FHandle, Integer(MYSQL_PLUGIN_DIR), '.\plugin\'); 
https://github.com/HeidiSQL/HeidiSQL/issues/163?timeline_page=1

**Comment by RvdHout** [comment]
Whoop, whoop... 🥳 

MySQLConnection.cpp

```
         // MariaDB Connector/C negotiates TLS by default and aborts the
         // connection if the server does not support it ("SSL is required, but
         // the server does not support it"). The old libmysql client did not.
         // Restore a "prefer TLS, fall back to plaintext" behavior: keep using
         // TLS when the server offers it, but do not require it and do not
         // verify the server certificate (matching the previous, unencrypted
         // default). These option enum values are MariaDB-specific, so only
         // apply them when the MariaDB connector is loaded.
         if (MySQLInterface::Instance()->IsMariaDB() && MySQLInterface::Instance()->p_mysql_options != 0)
         {
            const int HM_MYSQL_OPT_SSL_VERIFY_SERVER_CERT = 21;
            const int HM_MYSQL_OPT_SSL_ENFORCE = 38;
            const int HM_MYSQL_OPT_MYSQL_PLUGIN_DIR = 22;
            AnsiString sPluginDir = Unicode::ToANSI(MySQLInterface::Instance()->GetLibraryDirectory_() + _T("\\plugin"));

            char disabled = 0;
            MySQLInterface::Instance()->p_mysql_options(dbconn_, HM_MYSQL_OPT_SSL_ENFORCE, &disabled);
            MySQLInterface::Instance()->p_mysql_options(dbconn_, HM_MYSQL_OPT_SSL_VERIFY_SERVER_CERT, &disabled);
            MySQLInterface::Instance()->p_mysql_options(dbconn_, HM_MYSQL_OPT_MYSQL_PLUGIN_DIR, sPluginDir.c_str());
         }

```

i had to make String GetLibraryDirectory_(); public in MySQLInterface.h

**this is with:**
libmariadb.dll in `./Bin/` (libmariadb.dll 3.4.9)
caching_sha2_password.dll is in `./Bin/plugin/`

**and without** installing mariadb-connector-c-3.4.9-win64.msi

**Comment by RvdHout** [comment]
In the meanwhile tested libmariadb.dll with plugins against with MySQL 26.7.0, MySQL 8.0.24, MariaDB 10.11.18
All working OK

@martinknafve do you have something like: `C:\Program Files\MariaDB\MariaDB Connector C 64-bit\lib\plugin\` in your Windows PATH Environment Variables? As that would explain why you had no issues during your tests

**Comment by martinknafve** [comment]
Yeah, you are right. I tested in a clean VM now. When I use libmysql.dll then I can connect without any additional plugins, but when I use libmariadb.dll then I get the same error. It must have picked it up from a path when I tested it previously.

Maybe would be best to just bundle a recent build of libmysql.dll with the installer, to get rid of this type of issue. What do you think?

**Comment by RvdHout** [comment]
What about compatibility? Can a recent version of libmysql.dll still connect to older MySQL Installations?
libmariadb.dll seems more versatile, Support for MySQL 5.x, 8.x, 9x, 26.x as well as MariaDB 10.5, 10.6, 10.11 (and up most likely)

Besides libmariadb.dll usage can be easily fixed by including the plugins, i think you have seen it but if not [here](https://github.com/hmailserver/hmailserver/issues/479#issuecomment-5401669013)

**Comment by martinknafve** [comment]
Yeah, maybe using MariaDb would be better. I will upgrade my test environment to latest MySQL later this week and then run some tests.

**Comment by martinknafve** [comment]
Closing this. MariaDb client is now bundled with the installer, so there shouldn't be a need to manually copy any files to be able to connect to MySQL.
